src/Client.ts

```diff
@@ -344,53 +356,88 @@ export default class Client extends EventEmitter {
+        const transportEndpoints: Array<[Transport, TransportConfig]> = [];
+        let endpoints: { [key: string]: string[] } = {};
+        try {
+            endpoints = await (this as unknown as Agent).discoverBindings(
+                this.config.server!
+            );
+        } catch (e) {
+            console.error(e);
+        }
         for (const name of transportPref) {
-            let conf = this.config.transports![name];
-            if (!conf) {
+            const settings = this.config.transports![name];
+            if (!settings || !this.transports![name]) {
                 continue;
             }
-            if (typeof conf === 'string') {
-                conf = { url: conf };
-            } else if (conf === true) {
-                if (!endpoints) {
-                    try {
-                        endpoints = await ((this as unknown) as Agent).discoverBindings(
-                            this.config.server!
-                        );
-                    } catch (err) {
-                        console.error(err);
-                        continue;
-                    }
-                }
-                endpoints[name] = (endpoints[name] || []).filter(
-                    url => url.startsWith('wss:') || url.startsWith('https:')
-                );
-                if (!endpoints[name] || !endpoints[name].length) {
-                    continue;
-                }
-                conf = { url: endpoints[name][0] };
-            }
-
-            this.transport = new this.transports[name](
-                (this as unknown) as Agent,
+            const transport = new this.transports![name](
+                this as unknown as Agent,
                 this.sm,
                 this.stanzas
             );
-            this.transport.connect({
-                acceptLanguages: this.config.acceptLanguages || ['en'],
+
+            const config: TransportConfig = {
+                acceptLanguages: this.config.acceptLanguages || [this.config.lang ?? 'en'],
                 jid: this.config.jid!,
-                lang: this.config.lang || 'en',
+                lang: this.config.lang ?? 'en',
                 server: this.config.server!,
-                url: conf.url!,
-                ...conf
-            });
-            return;
+            };
+
+            if (typeof settings === 'string') {
+                transportEndpoints.push([transport, { ...config, url: settings }]);
+            } else if (settings === true) {
+                if (transport.discoverBindings) {
+                    for (const ep of (await transport.discoverBindings(this.config.server!) ?? [])) {
+                        transportEndpoints.push([transport, { ...config, ...ep }]);
+                    }
+                } else {
+                    for (const ep of (endpoints[name] ?? [])) {
+                        transportEndpoints.push([transport, { ...config, url: ep }]);
+                    }
```

Comment:
So Agent can `discoverBindings()`, but Transports can do so too? What's the difference between the found endpoints?

Reply:
Well BOSH and WebSocket both use the `.well-known/host-meta` and `.well-known/host-meta.json` files so there's no need to fetch them twice; `Agent::discoverBindings()` is defined in `plugins/hostmeta.ts` and will find endpoints for both of these before the first loop is run. `Transport::discoverBindings()` is for any transports that would need to do something specific, like TCP.

Reply:
A bit confusing, maybe renaming the `Agent` function and putting in a `NOTE` would be appropriate action.